### PR TITLE
Convert client factory methods to constructors

### PR DESCRIPTION
**Pull request by abhipsaMisra**

### CODE REVIEW

File: iothub/device/src/Edge/EdgeModuleClientHelper.cs [hunk 1]
```diff
@@ -31,29 +31,7 @@ internal class EdgeModuleClientFactory
...
-        /// <summary>
-        /// Creates an instance based on environment.
-        /// </summary>
-        public Task<IotHubModuleClient> CreateAsync()
-        {
-            return CreateInternalClientFromEnvironmentAsync();
-        }
-
-        private async Task<IotHubModuleClient> CreateInternalClientFromEnvironmentAsync()
+        internal static IotHubConnectionCredentials CreateIotHubConnectionCredentialsFromEnvironment()
```
Comment (abhipsaMisra) [inline iothub/device/src/Edge/EdgeModuleClientHelper.cs:34]: I'm looking at cleaning this class up a bit further, but I've opened the PR up for comments in the meanwhile.

File: e2e/test/helpers/TestDevice.cs [hunk 2]
```diff
@@ -175,18 +175,18 @@ public IotHubDeviceClient CreateDeviceClient(IotHubClientOptions options = defau
             {
                 if (authScope == ConnectionStringAuthScope.Device)
                 {
-                    deviceClient = IotHubDeviceClient.CreateFromConnectionString(ConnectionString, options);
+                    deviceClient = new IotHubDeviceClient(ConnectionString, options);
```
Comment (drwill-ms) [inline e2e/test/helpers/TestDevice.cs:178]: **gasp**

File: iothub/device/src/Authentication/IotHubConnectionCredentials.cs [hunk 3]
```diff
@@ -25,6 +26,9 @@ public sealed class IotHubConnectionCredentials : IConnectionCredentials
         /// <exception cref="ArgumentException"><paramref name="iotHubHostName"/> or device Id are an empty string or consist only of white-space characters.</exception>
         /// <exception cref="ArgumentException">Neither shared access key, shared access signature or X509 certificates were presented for authentication.</exception>
         /// <exception cref="ArgumentException">Either shared access key or shared access signature where presented together with X509 certificates for authentication.</exception>
+        /// <exception cref="ArgumentException"><see cref="DeviceAuthenticationWithX509Certificate"/> is used but <see cref="DeviceAuthenticationWithX509Certificate.Certificate"/> is null.</exception>
```
Comment (drwill-ms) [inline iothub/device/src/Authentication/IotHubConnectionCredentials.cs:29]: Can that exception be thrown when initalizing `DeviceAuthenticationWithX509Certificate` instead?

File: iothub/device/src/Authentication/IotHubConnectionCredentials.cs [hunk 4]
```diff
@@ -25,6 +26,9 @@ public sealed class IotHubConnectionCredentials : IConnectionCredentials
...
+        /// <exception cref="ArgumentException"><see cref="DeviceAuthenticationWithX509Certificate.ChainCertificates"/> is used over a protocol other than MQTT over TCP or AMQP over TCP></exception>
```
Comment (drwill-ms) [inline iothub/device/src/Authentication/IotHubConnectionCredentials.cs:30]:
```suggestion
        /// <exception cref="ArgumentException"><see cref="DeviceAuthenticationWithX509Certificate.ChainCertificates"/> is used over a protocol other than MQTT over TCP or AMQP over TCP.</exception>
```

File: iothub/device/src/Authentication/IotHubConnectionCredentials.cs [hunk 5]
```diff
@@ -342,6 +346,32 @@ private void Validate()
...
                     "Should not specify either SharedAccessKey or SharedAccessSignature if X.509 certificate is used for authenticating the client with IoT hub.");
             }
+
+            // Validate certs.
+            if (AuthenticationMethod is DeviceAuthenticationWithX509Certificate)
+            {
+                // Prep for certificate auth.
+                if (Certificate == null)
+                {
+                    throw new ArgumentException("No certificate was found. To use certificate authentication certificate must be present.");
```
Comment (drwill-ms) [inline iothub/device/src/Authentication/IotHubConnectionCredentials.cs:356]: When throwing ArgumentException and ArgumentNullException, let's be sure to specify the parameter name.
```suggestion
                    throw new ArgumentException("No certificate was found. To use certificate authentication certificate must be present.", nameof(authenticationMethod));
```

File: iothub/device/src/Edge/EdgeModuleClientHelper.cs [hunk 6]
```diff
@@ -103,27 +71,65 @@ private async Task<IotHubModuleClient> CreateInternalClientFromEnvironmentAsync(
...
 
-                ICertificateValidator certificateValidator = NullCertificateValidator.Instance;
+                return new IotHubConnectionCredentials(authMethod, hostName, gateway);
+            }
+        }
+
+        internal static async Task<ICertificateValidator> CreateCertificateValidatorFromEnvironmentAsync(ITrustBundleProvider trustBundleProvider, IotHubClientOptions options)
+        {
+            // Make sure client options is initialized.
+            if (options == default)
```
Comment (drwill-ms) [inline iothub/device/src/Edge/EdgeModuleClientHelper.cs:81]: Can we assume that by now options will have an instance? I'd be fine with a `Debug.Assert()` too.

File: iothub/device/src/Edge/EdgeModuleClientHelper.cs [hunk 7]
```diff
@@ -103,27 +71,65 @@ private async Task<IotHubModuleClient> CreateInternalClientFromEnvironmentAsync(
...
+                {
+                    Debug.WriteLine("EdgeModuleClientFactory setupTrustBundle from file");
+                    var expectedRoot = new X509Certificate2(certPath);
+                    certificateValidator = GetCertificateValidator(new List<X509Certificate2>() { expectedRoot }, options);
+                }
+
+                return certificateValidator;
+            }
+            else
+            {
```
Comment (drwill-ms) [inline iothub/device/src/Edge/EdgeModuleClientHelper.cs:108]: Else case not necessary, as previous block returns.

File: iothub/device/src/IotHubDeviceClient.cs [hunk 8]
```diff
@@ -25,57 +25,73 @@ public class IotHubDeviceClient : IDisposable
         /// </summary>
         public const uint DefaultOperationTimeoutInMilliseconds = 4 * 60 * 1000;
 
-        private IotHubDeviceClient(InternalClient internalClient)
+        /// <summary>
+        /// Creates a disposable <c>IotHubDeviceClient</c> from the specified connection string.
+        /// </summary>
+        /// <param name="connectionString">The connection string based on shared access key used in API calls which allows the device to communicate with IoT Hub.</param>
+        /// <param name="options">The options that allow configuration of the device client instance during initialization.</param>
+        /// <returns>A disposable <c>IotHubDeviceClient</c> instance.</returns>
```
Comment (drwill-ms) [inline iothub/device/src/IotHubDeviceClient.cs:33]:
```suggestion
        /// <returns>A disposable client instance.</returns>
```

File: iothub/device/src/IotHubDeviceClient.cs [hunk 9]
```diff
@@ -25,57 +25,73 @@ public class IotHubDeviceClient : IDisposable
...
+        /// <exception cref="ArgumentNullException"><paramref name="connectionString"/>, IoT hub host name or device Id is null.</exception>
```
Comment (drwill-ms) [inline iothub/device/src/IotHubDeviceClient.cs:34]:
```suggestion
        /// <exception cref="ArgumentNullException">In the <paramref name="connectionString"/>, the IoT hub host name or device Id is null.</exception>
```

File: iothub/device/src/IotHubDeviceClient.cs [hunk 10]
```diff
@@ -25,57 +25,73 @@ public class IotHubDeviceClient : IDisposable
...
-                throw new ArgumentException("A module Id was specified in the connection string - please use IotHubModuleClient for modules.");
-            }
-
-            if (Logging.IsEnabled)
-                Logging.Associate(this, this, internalClient, nameof(IotHubDeviceClient));
         }
 
         /// <summary>
-        /// Creates a disposable DeviceClient from the specified parameters, that uses AMQP transport protocol.
+        /// Creates a disposable <c>IotHubDeviceClient</c> from the specified parameters.
```
Comment (drwill-ms) [inline iothub/device/src/IotHubDeviceClient.cs:44]:
```suggestion
        /// Creates a disposable instance of this client.
```

File: iothub/device/src/IotHubDeviceClient.cs [hunk 11]
```diff
@@ -25,57 +25,73 @@ public class IotHubDeviceClient : IDisposable
...
+                }
+            }
+
+            // Make sure client options is initialized.
+            if (options == default)
+            {
+                options = new();
+            }
+
+            InternalClient = new InternalClient(iotHubConnectionCredentials, options, null);
```
Comment (drwill-ms) [inline iothub/device/src/IotHubDeviceClient.cs:88]: Next step is to make this a base class?

File: iothub/device/tests/ClientSettings/TransportSettingsTests.cs [hunk 12]
```diff
@@ -191,7 +191,7 @@ public void IotHubDeviceClient_NullX509Certificate()
             var options = new IotHubClientOptions(new IotHubClientAmqpSettings { PrefetchCount = 100 });
 
             // act
-            _ = IotHubDeviceClient.Create(hostName, authMethod, options);
+            _ = new IotHubDeviceClient(hostName, authMethod, options);
```
Comment (drwill-ms) [inline iothub/device/tests/ClientSettings/TransportSettingsTests.cs:194]: needs disposal

File: iothub/device/tests/Edge/EdgeModuleClientHelperTest.cs [hunk 13]
```diff
@@ -181,8 +160,11 @@ public async Task TestCreate_FromEnvironment_ShouldCreateClient()
             Environment.SetEnvironmentVariable(AuthSchemeVariableName, "sasToken");
 
             var settings = new IotHubClientAmqpSettings();
+            var options = new IotHubClientOptions(settings);
             var trustBundle = Substitute.For<ITrustBundleProvider>();
-            IotHubModuleClient dc = await new EdgeModuleClientFactory(trustBundle).CreateAsync();
+            IotHubConnectionCredentials creds = EdgeModuleClientHelper.CreateIotHubConnectionCredentialsFromEnvironment();
+            ICertificateValidator certValidator = await EdgeModuleClientHelper.CreateCertificateValidatorFromEnvironmentAsync(trustBundle, options).ConfigureAwait(false);
+            IotHubModuleClient dc = new IotHubModuleClient(creds, options, certValidator);
```
Comment (drwill-ms) [inline iothub/device/tests/Edge/EdgeModuleClientHelperTest.cs:167]: using?

File: iothub/device/tests/IotHubDeviceClientTests.cs [hunk 14]
```diff
@@ -952,7 +952,7 @@ public void DeviceClientOnConnectionClosedInvokeHandlerAndRecoveryForStatusChang
         [TestMethod]
         public void CompleteAsyncThrowsForNullMessage()
         {
-            IotHubDeviceClient client = IotHubDeviceClient.CreateFromConnectionString(FakeConnectionString);
+            IotHubDeviceClient client = new IotHubDeviceClient(FakeConnectionString);
```
Comment (drwill-ms) [inline iothub/device/tests/IotHubDeviceClientTests.cs:955]: some of these are missing the `using` keyword

File: iothub/device/tests/ModuleClientTests.cs [hunk 15]
```diff
@@ -51,27 +51,27 @@ public class ModuleClientTests
...
             Assert.IsNotNull(moduleClient);
         }
 
         [TestMethod]
-        [ExpectedException(typeof(ArgumentException))]
+        [ExpectedException(typeof(ArgumentNullException))]
         public void ModuleClient_CreateFromConnectionString_WithNoModuleIdThrows()
         {
-            IotHubModuleClient.CreateFromConnectionString(ConnectionStringWithoutModuleId);
+            new IotHubModuleClient(ConnectionStringWithoutModuleId);
```
Comment (drwill-ms) [inline iothub/device/tests/ModuleClientTests.cs:68]:
```suggestion
            _ = new IotHubModuleClient(ConnectionStringWithoutModuleId);
```

File: iothub/device/tests/ModuleClientTests.cs [hunk 16]
```diff
@@ -51,27 +51,27 @@ public class ModuleClientTests
...
         }
 
         [TestMethod]
         public void ModuleClient_CreateFromConnectionString_NoTransportSettings()
         {
-            var moduleClient = IotHubModuleClient.CreateFromConnectionString(FakeConnectionString);
+            var moduleClient = new IotHubModuleClient(FakeConnectionString);
```
Comment (drwill-ms) [inline iothub/device/tests/ModuleClientTests.cs:74]: More missing `using` keywords in this file too.

File: iothub/device/src/Edge/EdgeModuleClientHelper.cs [hunk 17]
```diff
@@ -103,27 +71,65 @@ private async Task<IotHubModuleClient> CreateInternalClientFromEnvironmentAsync(
...
+                    certificateValidator = GetCertificateValidator(certs, options);
                 }
 
-                return new IotHubModuleClient(CreateInternalClientFromAuthenticationMethod(hostname, authMethod, _options), certificateValidator);
+                return certificateValidator;
             }
         }
 
-        private ICertificateValidator GetCertificateValidator(IList<X509Certificate2> certs)
+        private static ICertificateValidator GetCertificateValidator(IList<X509Certificate2> certs, IotHubClientOptions options)
```
Comment (timtay-microsoft) [inline iothub/device/src/Edge/EdgeModuleClientHelper.cs:124]:
```suggestion
        private static ICertificateValidator CreateCertificateValidator(IList<X509Certificate2> certs, IotHubClientOptions options)
```

nit: it is being created, not retrieved

Comment (timtay-microsoft) [reply iothub/device/src/Edge/EdgeModuleClientHelper.cs:124]: static "getters" are a bit dubious, but static "creators" are safe

**Review by timtay-microsoft — APPROVED** [review]

File: iothub/device/src/Authentication/IotHubConnectionCredentials.cs (hunk 3 above)
Comment (abhipsaMisra) [reply iothub/device/src/Authentication/IotHubConnectionCredentials.cs:29]: That's a good idea, yeah

File: iothub/device/src/IotHubDeviceClient.cs (hunk 9 above)
Comment (abhipsaMisra) [reply iothub/device/src/IotHubDeviceClient.cs:34]: reworded

File: iothub/device/src/IotHubDeviceClient.cs (hunk 11 above)
Comment (abhipsaMisra) [reply iothub/device/src/IotHubDeviceClient.cs:89]: That's a good idea. We should add this to our task list.

File: e2e/test/iothub/messaging/MessageReceiveFaultInjectionTests.cs [hunk 18]
```diff
@@ -397,8 +397,6 @@ async Task InitOperationAsync(IotHubDeviceClient deviceClient, TestDevice testDe
                         // ignore and proceed.
                     }
                 }
-
-                await serviceClient.Messaging.CloseAsync().ConfigureAwait(false);
```
Comment (abhipsaMisra) [inline e2e/test/iothub/messaging/MessageReceiveFaultInjectionTests.cs:401]: @timtay-microsoft I see that this was added in #2614 . The service client instance in this test is used throughout the lifetime of the test, so I've removed this "close" call from the "InitOperationAsync" task and left it only in the "CleanupOperationAsync" task. Let me know if there is some other reason why this is required in the init operation block.

**Comment by abhipsaMisra** [comment]
The test failures here are from the complete E2E test run:
MessageFeedbackReceiver_Operation - Expected messagedFeedbackReceived to be true because Timed out waiting to receive message feedback., but found False.
DevicesClient_ImportDevices, DevicesClient_ExportDevices - Microsoft.Azure.Devices.Common.Exceptions.QuotaExceededException

**Review by drwill-ms — APPROVED** [review]

**Review by drwill-ms — APPROVED** [review]

**Pull request merged**